draft-ietf-quic-qpack.md

```diff
@@ -52,46 +52,162 @@ code and issues list for this draft can be found at
+the decoder, the dynamic table size is strictly bounded.
+
+The decoder determines the maximum size that the encoder is permitted to use for
+the dynamic table.  In HTTP/QUIC, this value is determined by the
+SETTINGS_HEADER_TABLE_SIZE setting (see Section 4.2.5.2 of {{!QUIC-HTTP}}).
+
+An encoder can choose to use less capacity than this maximum size (see
+{{size-update}}), but the chosen size MUST stay lower than or equal to the
+maximum set by the decoder.  Whenever the maximum size for the dynamic table is
+reduced, entries are evicted from the end of the dynamic table until the size of
+the dynamic table is less than or equal to the maximum size.
+
+This mechanism can be used to completely clear entries from the dynamic table by
+setting a maximum size of 0, which can subsequently be restored.
+
+### Indexing
+
+Entries in the QPACK static and dynamic tables are addressed separately.
+
+While entries in the static table have the same indices at all times, entries
+are inserted into the dynamic table over time.  Each entry possesses both an
+absolute index which is fixed for the lifetime of that entry and a relative
+index which changes over time based on the context of the reference. The first
+entry inserted has an absolute index of "1"; indices increase sequentially with
+each insertion.
+
+On the control stream, a relative index of "0" always refers to the most
+recently inserted value in the dynamic table.  Note that this means the
+entry referenced by a given relative index can change while interpreting
+a HEADERS frame as new entries are inserted.
+
+~~~~~ drawing
+    +---+---------------+-------+
+    | n |      ...      | d + 1 |  Absolute Index
+    + - +---------------+   -   +
+    | 0 |      ...      | n-d-1 |  Relative Index
+    +---+---------------+-------+
+      ^                     |
+      |                     V
+Insertion Point         Dropping Point
+
+n = count of entries inserted
+d = count of entries dropped
+~~~~~
+{: title="Example Dynamic Table Indexing - Control Stream"}
+
+Because frames from request streams can be delivered out of order with
+instructions on the control stream, relative indices are relative to the Base
+Index at the beginning of the header block (see {{absolute-index}}). The Base
+Index is the absolute index of the entry which has the relative index of zero
+when interpreting the frame.  The relative indices of entries do not change
+while interpreting headers on a request or push stream.
+
+~~~~~ drawing
+             Base Index
+                 |
+                 V
+    +---+-----+-----+-----+-------+
+    | n | n-1 | n-2 | ... |  d+1  |  Absolute Index
+    +---+-----+  -  +-----+   -   +
+              |  0  | ... | n-d-3 |  Relative Index
+              +-----+-----+-------+
+
+n = count of entries inserted
+d = count of entries dropped
+~~~~~
+{: title="Example Dynamic Table Indexing - Request Stream"}
+
+Entries with an absolute index greater than a frame's Base Index can be
```

Comment:
Explain why: "A header block on the request stream can reference entries added after the entry identified by the Base Index.  This allows an encoder to process a header block in a single pass and include references to entries added while processing this (or other) header blocks.  Newly added entries are referenced using Post-Base instructions.  Indices for Post-Base instructions increase in the same direction as absolute indices, but the zero value is one higher than the Base Index."

Reply:
Also, is the Largest Reference a Post-Base index?  It should be.

Reply:
Not quite, though LR and Base are given one as absolute and one relative to the other. LR might be less than Base, or more, or the same.

Reply:
I don't think that there is any value in having Largest Reference with a lower absolute value than the Base Index.  I mean, sure, it can happen, but why would you encode that?

Reply:
Because a likely algorithm would be that Base is the largest entry you've inserted in the dynamic table, before you've examined the incoming header set.  If you don't wind up using entries that are in flight (or any entries, in the extreme case), then Largest Reference will indicate that the most recent entries aren't needed.

Reply:
Right.  Of course.